### PR TITLE
Button Refactor Base

### PR DESCRIPTION
**Pull request by rb114-hw**
### Additions
**ButtonBase.tsx**
- Util file that can be used to create buttons. Currently limited uses but has the ability to expand

### Changes
**DummyPost.tsx**
- Implemented Button Base (with onClickNav function)

**EditProfile.tsx**
- Implemented Button Base

**Profile.tsx**
- Implemented Button Base (with onClickNav function)

**CreatePost.tsx**
- Implemented Button Base

**FAQ.tsx**
- Implemented Button Base (with onClickNav function)

**ManageAccount.tsx**
- Implemented Button Base (with onClickNav function)

**Notifications.tsx**
- Implemented Button Base (with onClickNav function)

**Settings.tsx**
- Implemented Button Base (with onClickNav function)

**ImagePicker.tsx**
- Implemented Button Base
- Left some pre-existing buttons as comments as the current button base is unable to implement the spinner

### CODE REVIEW

**Review by wrt95 — CHANGES_REQUESTED** [review]
@rb114-hw Can you upload photos of how it looks in the different pages?

**Comment by rb114-hw** [comment]
### Screenshots
**Create Post Page**
Defualt
image
Selected Image
image

**Profile Page**
Default
image

**Edit Profile Page**
Defualt
image
Image Selected
image

**Post Page**
Defualt
image

**Settings Page**
Defualt
image

**Manage Acount Page**
Defualt
image

**Notifications Page**
Defualt
image

**FAQ Page**
Defualt
image

**Comment by wrt95** [comment]
@rb114-hw 
Can we make the cancel changes and cancel post buttons small, but red? 
So for example adding a new property where you specify the colour as a prop for the button? 

Ill comment in the code so you can see.

File: src/Components/Profile/Logged-in-user/Profile.tsx [hunk 1]
```diff
@@ -105,23 +112,8 @@ const Profile = (props: ProfileProps) => {
...
-                    >
-                        <Icon
-                            name="edit"
-                            type="FontAwesome5"
-                            style={styles.button_icon}
-                        />
-                        <Text style={styles.button_text}>Edit Profile</Text>
-                    </Button>
+                <View style={styles.edit_container}>
+                    <ButtonBase buttonText="Edit Profile" icon="edit" onClick={onClickNavigate} buttonType="small" />
```
Comment (wrt95) [inline src/Components/Profile/Logged-in-user/Profile.tsx:116]: <ButtonBase buttonText="Edit Profile" icon="edit" onClick={onClickNavigate} buttonType="small" colour="green" />

File: src/Components/create-post/CreatePost.tsx [hunk 2]
```diff
@@ -234,26 +235,11 @@ const CreatePost = (props: CreatePostProps) => {
...
-                        style={styles.create_button}
-                    >
-                        <Icon
-                            name="save"
-                            type="FontAwesome5"
-                            style={styles.button_icon}
-                        />
-                        <Text style={styles.button_text}>Create</Text>
-                    </Button>
+                    <ButtonBase buttonText="Create" icon="save" onClick={checkValidPost} buttonType="small" />
```
Comment (wrt95) [inline src/Components/create-post/CreatePost.tsx:238]: <ButtonBase buttonText="Create" icon="save" onClick={checkValidPost} buttonType="small" colour="green" />

File: src/Components/create-post/CreatePost.tsx [hunk 3]
```diff
@@ -234,26 +235,11 @@ const CreatePost = (props: CreatePostProps) => {
...
                 </View>
 
                 <View>
-                    <Button
-                        onPress={() => doCancel()}
-                        style={styles.cancel_button}
-                    >
-                        <Text style={styles.button_text}>Cancel</Text>
-                    </Button>
+                    <ButtonBase buttonText="Cancel Post" onClick={doCancel} buttonType="cancel" />
```
Comment (wrt95) [inline src/Components/create-post/CreatePost.tsx:242]: <ButtonBase buttonText="Cancel Post" onClick={doCancel} buttonType="small" colour="red" />

File: src/Components/settings/FAQ/FAQ.tsx [hunk 4]
```diff
@@ -41,19 +42,7 @@ const FAQ = (props: FAQProps) => {
...
-                            props.navigation.navigate("Settings");
-                            props.navigation.dangerouslyGetParent().setOptions({
-                                tabBarVisible: true
-                            });
-                        }}
-                    >
-                        <Text style={styles.button_text}>Return to Settings</Text>
-                    </Button>
-                </View>
+                <ButtonBase buttonText="Return to Settings" onClick={onClickReturn} buttonType="cancel" />
```
Comment (wrt95) [inline src/Components/settings/FAQ/FAQ.tsx:45]: <ButtonBase buttonText="Return to Settings" onClick={onClickReturn} buttonType="large" colour="red" />

File: src/Components/settings/ManageAccount.tsx [hunk 5]
```diff
@@ -201,19 +209,7 @@ const ManageAccount = (props: ManageAccountProps) => {
...
-                            props.navigation.navigate("Settings");
-                            props.navigation.dangerouslyGetParent().setOptions({
-                                tabBarVisible: true
-                            });
-                        }}
-                    >
-                        <Text style={styles.button_text}>Return to Settings</Text>
-                    </Button>
-                </View>
+                <ButtonBase buttonText="Return to Settings" onClick={onClickReturn} buttonType="cancel" />
```
Comment (wrt95) [inline src/Components/settings/ManageAccount.tsx:212]: <ButtonBase buttonText="Return to Settings" onClick={onClickReturn} buttonType="large" colour="red" />

File: src/Components/settings/Notifications/Notifications.tsx [hunk 6]
```diff
@@ -83,19 +84,7 @@ const Notifications = (props: NotificationsProps) => {
...
-                            props.navigation.navigate("Settings");
-                            props.navigation.dangerouslyGetParent().setOptions({
-                                tabBarVisible: true
-                            });
-                        }}
-                    >
-                        <Text style={styles.button_text}>Return to Settings</Text>
-                    </Button>
-                </View>
+                <ButtonBase buttonText="Return to Settings" onClick={onClickReturn} buttonType="cancel" />
```
Comment (wrt95) [inline src/Components/settings/Notifications/Notifications.tsx:87]: <ButtonBase buttonText="Return to Settings" onClick={onClickReturn} buttonType="large" colour="red" />

File: src/Components/settings/Settings.tsx [hunk 7]
```diff
@@ -106,19 +115,9 @@ const Settings = (props: SettingsProps) => {
...
-                    onPress={() => {
-                        props.navigation.navigate("Login")
-                    }}
-                >
-                    <Text style={styles.button_text}>
-                        Log Out
-                    </Text>
-                </Button>
-            </View>
+            <ButtonBase buttonText="Log Out" onClick={onClickReturn} buttonType="cancel" />
```
Comment (wrt95) [inline src/Components/settings/Settings.tsx:118]: <ButtonBase buttonText="Log Out" onClick={onClickReturn} buttonType="large" colour="red" />

File: src/Components/util/ButtonBase.tsx [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+} from "native-base";
+
+/**
+ * Props for the Button component
+ */
+interface ButtonBaseProps {
+    buttonText: string;
+    icon?: string;
+    removeTabBar?: boolean;
+    buttonType: "small" | "cancel" | "regular";
```
Comment (wrt95) [inline src/Components/util/ButtonBase.tsx:19]: change to: 
buttonType: "small" | "large";

File: src/Components/util/ButtonBase.tsx [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+/**
+ * Component that displays an error and text passed in
+ * via props.
+ *
+ * @param props     The props of this component
+ */
+export const ButtonBase = (props: ButtonBaseProps) => {
+    /** DISPLAY ICON IF IT EXISTS **/
+    const iconPicker = () => {
```
Comment (wrt95) [inline src/Components/util/ButtonBase.tsx:31]: Can simplify this function to: 

```
    const iconPicker = () => {
        return props.icon ?
            <Icon
                name={props.icon}
                type="FontAwesome5"
                style={styles.button_icon}
        /> : <></>
```

File: src/Components/util/ButtonBase.tsx [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            )
+        }
+    }
+    
+    const buttonPicker = () => {
+        switch(props.buttonType) {
+            case "small":
+                return (
+                    <Button
+                        style={styles.small_button}
```
Comment (wrt95) [inline src/Components/util/ButtonBase.tsx:51]: style={props.colour === "green" ? styles.small_button_green : styles.small_button_red}

File: src/Components/util/ButtonBase.tsx [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                        onPress={props.onClick}
+                    >
+                        <Text style={styles.button_text}>{props.buttonText}</Text>
+                        {iconPicker()}
+                    </Button>
+                )
+            case "cancel":
```
Comment (wrt95) [inline src/Components/util/ButtonBase.tsx:58]: case "large":

File: src/Components/util/ButtonBase.tsx [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+                return (
+                    <Button
+                        style={styles.cancel_button}
```
Comment (wrt95) [inline src/Components/util/ButtonBase.tsx:61]: style={props.colour === "green" ? styles.large_button_green : styles.large_button_red}

File: src/Components/util/ButtonBase.tsx [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                        onPress={props.onClick}
+                    >
+                        <Text style={styles.cancel_button_text}>{props.buttonText}</Text>
+                        {iconPicker()}
+                    </Button>
+                )
+            case "regular":
```
Comment (wrt95) [inline src/Components/util/ButtonBase.tsx:68]: delete

File: src/Components/util/ButtonBase.tsx [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+const styles = StyleSheet.create({
+    cancel_button: {
+        backgroundColor: '#cc0000',
+        width: 320,
+        marginTop: 40,
+        marginBottom: 20,
+        borderRadius: 10,
+        justifyContent: 'center',
+    },
+    small_button: {
```
Comment (wrt95) [inline src/Components/util/ButtonBase.tsx:97]: small_button_red: {}
small_button_green: {}

File: src/Components/util/ButtonBase.tsx [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+
+    return (
+        <View>
+            {buttonPicker()}
+        </View>
+    );
+}
+
+const styles = StyleSheet.create({
+    cancel_button: {
```
Comment (wrt95) [inline src/Components/util/ButtonBase.tsx:89]: large_button_red: {}
large_button_green: {}

File: src/Components/util/ImagePicker.tsx [hunk 16]
```diff
@@ -135,17 +135,7 @@ const ImagePickerr = (props: ImagePickerProps) => {
...
-                        style={styles.cancel_button}
-                    >
-                        <Icon
-                            name="times"
-                            type="FontAwesome5"
-                            style={styles.button_icon}
-                        />
-                        <Text style={styles.button_Text}>Remove Image</Text>
-                    </Button>
+                    <ButtonBase buttonText="Remove Image" icon="times" onClick={removeImage} buttonType="cancel" />
```
Comment (wrt95) [inline src/Components/util/ImagePicker.tsx:138]: <ButtonBase buttonText="Remove Image" icon="times" onClick={removeImage} buttonType="large" colour="red" />

File: src/Components/util/ImagePicker.tsx [hunk 17]
```diff
@@ -173,20 +163,13 @@ const ImagePickerr = (props: ImagePickerProps) => {
...
-                        <Icon
-                            name="camera"
-                            type="FontAwesome5"
-                            style={styles.button_icon}
-                        />
-                        <Text style={styles.button_Text}>Take Photo</Text>
-                    </Button>
-
+                <View style={props.imageSelected ? { display: "none" } : styles.blank}>
+                    <ButtonBase buttonText="Take Photo" icon="camera" onClick={launchCameraOnPhone} buttonType="regular" />
```
Comment (wrt95) [inline src/Components/util/ImagePicker.tsx:167]: <ButtonBase buttonText="Take Photo" icon="camera" onClick={launchCameraOnPhone} buttonType="large" colour="green" />

File: src/Components/util/ImagePicker.tsx [hunk 18]
```diff
@@ -173,20 +163,13 @@ const ImagePickerr = (props: ImagePickerProps) => {
...
+                    
                     {/** GET IMAGES FROM CAMERAROLL **/}
+                    <ButtonBase buttonText="Choose from Library" icon="images" onClick={launchLibraryFromPhone} buttonType="regular" />
```
Comment (wrt95) [inline src/Components/util/ImagePicker.tsx:170]: <ButtonBase buttonText="Choose from Library" icon="images" onClick={launchLibraryFromPhone} buttonType="large" colour="green" />

File: src/Components/util/ButtonBase.tsx [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        backgroundColor: '#cc0000',
+        width: 320,
```
Comment (wrt95) [inline src/Components/util/ButtonBase.tsx:91]: width: 300

**Review by wrt95 — CHANGES_REQUESTED** [review]
This should be it i think! 
Good job, but a little more has to be done

**Comment by rb114-hw** [comment]
Suggestions Implemented! I'll need to sort out the css elements of the large/small buttons. Then everything should fall into place

**Comment by rb114-hw** [comment]
### Updated Screenshots
**Create Post Page**
Defualt
image
Image selected
image

**Profile Page**
Defualt
image

**Post Page**
Defualt
image

**Edit Profile Page**
Defualt
image
Image Selected
image

**Settings Page**
Defualt
image

**Review by calumshep — CHANGES_REQUESTED** [review]
Not sure how up-to-date the photos in the most recent comment on this PR are, but I have a couple of suggestions:

- Make the buttons all the same width (so Save and Cancel buttons on create post page are the same width as the choose photo images for example)
- Reduce the font size within buttons to increase spacing around the text

**Comment by wrt95** [comment]
@calumshep 
Photos are not up to date. All sizes are fixed, and we have 3 types of buttons (small, large and small-rounded). It all looks very good. Save and Cancel are now the same size (see comments on Teams). 
Font size is also reduced.

**Comment by wrt95** [comment]
@calumshep 
Looks like this now, which I think is how we want it to be for now. 
Screenshot 2021-02-13 at 14 26 19
Screenshot 2021-02-13 at 14 26 14
Screenshot 2021-02-13 at 14 26 10
Screenshot 2021-02-13 at 14 26 06
Screenshot 2021-02-13 at 14 26 01
Screenshot 2021-02-13 at 14 25 54
Screenshot 2021-02-13 at 14 25 50
Screenshot 2021-02-13 at 14 25 42
Screenshot 2021-02-13 at 14 25 32

**Comment by calumshep** [comment]
Yep, mostly looks good now with the three consistent button types. Just one small thing - the text and icon inside the Create/Save/Edit Profile buttons is off-centre and I suspect it may be because the font-awesome icon has too much right-padding? Is this something you can fix?

**Review by sp103 — DISMISSED** [review]
all looks good to me

**Comment by wrt95** [comment]
@calumshep 

Good spotted! 
Actually, no paddingRight was involved. The reason of the not centring with icons was that the Icon didnt have a <Text> block around it. 

This code solved the issue: 
```<Text style={styles.icon_container}>{iconPicker()}</Text>```

Photos: 
Screenshot 2021-02-14 at 14 23 41
Screenshot 2021-02-14 at 14 23 20

**Review by calumshep — APPROVED** [review]

**Review by wrt95 — APPROVED** [review]

**Pull request merged**